docs(compute-samples): added samples and tests for custom machine type

Sita04 opened the pull request:
Added Compute Machine type samples and tests.
Docs: https://cloud.google.com/compute/docs/instances/creating-instance-with-custom-machine-type

Sita04 commented:
Kindly provision the required permission: roles/compute.instanceAdmin.v1 and roles/compute.storageAdmin
TIA!

Sita04 commented on compute/cloud-client/src/main/java/compute/custommachinetype/CreateCustomMachineType.java:
```diff
@@ -0,0 +1,105 @@
+    // Project ID or project number of the Cloud project you want to use.
+    String projectId = "your-google-cloud-project-id";
+    // Name of the zone to create the instance in. For example: "us-west3-b".
+    String zone = "gcloud-zone";
```
Done.

Sita04 commented on compute/cloud-client/src/main/java/compute/custommachinetype/CreateCustomMachineType.java:
```diff
@@ -0,0 +1,105 @@
+              // Describe the size and source image of the boot disk to attach to the instance.
+              AttachedDiskInitializeParams.newBuilder()
+                  .setSourceImage(
+                      String.format("projects/debian-cloud/global/images/family/%s", "debian-11"))
```
I added a formatter, so that the customer will know that it's a configurable parameter and not overlook it by mistake.
wdyt?

Shabirmean replied:
Makes sense. Maybe then we need 2 formatting strings? 
`"projects/%s/global/images/family/%s", "debian-cloud", "debian-11"`

averikitsch replied:
Can you add a comment on where to find these values?

Sita04 replied:
Done. 
Included 2 formatted arguments and added the comment on how to find the public images.

Sita04 commented on compute/cloud-client/src/main/java/compute/custommachinetype/CreateCustomMachineType.java:
```diff
@@ -0,0 +1,105 @@
+          .get(3, TimeUnit.MINUTES);
+      // Check for errors.
+      if (response.hasError()) {
+        System.out.println("Instance creation failed ! ! " + response);
```
Not following any pattern. Modified it. Thanks.

Sita04 commented on compute/cloud-client/src/main/java/compute/custommachinetype/CreateInstanceWithCustomSharedCore.java:
```diff
@@ -0,0 +1,283 @@
+    private static final Map<String, CpuSeries> ENUM_MAP;
+
+    // Build an immutable map of String name to enum pairs.
+    static {
+      Map<String, CpuSeries> map = new ConcurrentHashMap<>();
+      for (CpuSeries instance : CpuSeries.values()) {
+        map.put(instance.getCpuSeries(), instance);
+      }
+      ENUM_MAP = Collections.unmodifiableMap(map);
+    }
```
I've modified the static block to this:
```    
    static {
      ENUM_MAP = init();
    }

    // Build an immutable map of String name to enum pairs.
    public static Map<String, CpuSeries> init() {
      Map<String, CpuSeries> map = new ConcurrentHashMap<>();
      for (CpuSeries instance : CpuSeries.values()) {
        map.put(instance.getCpuSeries(), instance);
      }
      return Collections.unmodifiableMap(map);
    }
```
PTAL.

Shabirmean replied:
This change looks good!

Sita04 commented on compute/cloud-client/src/main/java/compute/custommachinetype/CreateInstanceWithCustomSharedCore.java:
```diff
@@ -0,0 +1,283 @@
+    }
+  }
+
+  // The limits for various CPU types are described on:
```
Added a comment describing the enum and modified the typo.

Sita04 commented on compute/cloud-client/src/main/java/compute/custommachinetype/CreateInstanceWithCustomSharedCore.java:
```diff
@@ -0,0 +1,283 @@
+    }
+  }
+
+  static final class TypeLimits {
```
Done.

savijatv commented on compute/cloud-client/src/main/java/compute/custommachinetype/CreateWithHelper.java:
```diff
@@ -0,0 +1,287 @@
+          cpuSeries, typeLimit.extraMemoryLimit);
+    }
+
+    // Return the custom machine type in form of a string acceptable by Compute Engine API.
```
in the form of a string

savijatv commented on compute/cloud-client/src/main/java/compute/custommachinetype/CreateWithHelper.java:
```diff
@@ -0,0 +1,287 @@
+    return String.format("zones/%s/machineTypes/%s-%s-%s", zone, cpuSeries, coreCount, memory);
+  }
+
+  // Creates a new VM instance with a custom machine type.
```
Create a VM instance with a ...

savijatv commented on compute/cloud-client/src/main/java/compute/custommachinetype/CreateWithoutHelper.java:
```diff
@@ -0,0 +1,111 @@
+        coreCount, memory);
+  }
+
+  // Create new VM instances without using a CustomMachineType class.
```
Create a VM instance without ..

savijatv commented on compute/cloud-client/src/main/java/compute/custommachinetype/CreateWithoutHelper.java:
```diff
@@ -0,0 +1,111 @@
+          .setBoot(true)
+          .build();
+
+      // Collect information into the Instance object.
```
Same comment as above.

savijatv commented on compute/cloud-client/src/main/java/compute/custommachinetype/HelperClass.java:
```diff
@@ -0,0 +1,252 @@
+    return null;
+  }
+
+  // Return the custom machine type in form of a string acceptable by Compute Engine API.
```
in the form of a ...

savijatv commented on compute/cloud-client/src/main/java/compute/custommachinetype/HelperClass.java:
```diff
@@ -0,0 +1,252 @@
+    return machineType[machineType.length - 1];
+  }
+
+  // Construct a custom machine type.
```
Create a custom..

savijatv commented on compute/cloud-client/src/main/java/compute/custommachinetype/UpdateMemory.java:
```diff
@@ -0,0 +1,125 @@
+        }
+      }
+
+      // Modify the machine definition, remember that extended memory
```
Modify the machine definition. Note that extended memory....

savijatv commented on compute/cloud-client/src/main/java/compute/custommachinetype/UpdateMemory.java:
```diff
@@ -0,0 +1,125 @@
+      String machineType = instance.getMachineType();
+      String start = machineType.substring(0, machineType.lastIndexOf("-"));
+
+      // Prepare the request.
```
Create the machine type instance request object?

savijatv commented on compute/cloud-client/src/main/java/compute/custommachinetype/UpdateMemory.java:
```diff
@@ -0,0 +1,125 @@
+                  .build())
+              .build();
+
+      // Wait for the operation to complete.
```
Can we elaborate this operation a little bit?

savijatv commented on compute/cloud-client/src/test/java/compute/custommachinetype/CustomMachineTypeIT.java:
```diff
@@ -0,0 +1,187 @@
+    requireEnvVar("GOOGLE_APPLICATION_CREDENTIALS");
+    requireEnvVar("GOOGLE_CLOUD_PROJECT");
+
+    // Cleanup existing stale resources.
```
Clean up existing...

averikitsch commented on compute/cloud-client/src/main/java/compute/custommachinetype/CreateCustomMachineType.java:
```diff
@@ -0,0 +1,105 @@
+    // For example: "zones/europe-west3-c/machineTypes/f1-micro"
+    // OR
+    // It can be a CustomMachineType object, describing a custom type you want to use.
+    String machineType = "zones/{zone}/machineTypes/{typeName}";
```
Is there not a helper method for creating this string?

Sita04 replied:
Yes, there is a helper method.
But the setMachineType() accepts only a string and hence I thought this might be an easier approach, rather than creating a machineType and wrapping it to string.

averikitsch commented on compute/cloud-client/src/main/java/compute/custommachinetype/CreateCustomMachineType.java:
```diff
@@ -0,0 +1,105 @@
+      String instanceName, String machineType)
+      throws IOException, ExecutionException, InterruptedException, TimeoutException {
+
+    try (InstancesClient instancesClient = InstancesClient.create()) {
```
Please add comment on client initialization: [go/cloud-dpe-repo-onboarding](http://go/cloud-dpe-repo-onboarding)

Sita04 replied:
Done. Thanks.

averikitsch commented on compute/cloud-client/src/main/java/compute/custommachinetype/CreateCustomMachineType.java:
```diff
@@ -0,0 +1,105 @@
+      // Check for errors.
+      if (response.hasError()) {
+        System.out.println("Instance creation failed ! ! " + response);
+        return Instance.getDefaultInstance();
```
I agree. Unless we have a recommendation on how to handle a response error, then we should throw an error.

Sita04 replied:
Done. Refactored the snippet to not return anything.

averikitsch commented on compute/cloud-client/src/main/java/compute/custommachinetype/CreateInstanceWithCustomSharedCore.java:
```diff
@@ -0,0 +1,283 @@
+
+    if (!Arrays.asList(CpuSeries.E2_SMALL.cpuSeries, CpuSeries.E2_MICRO.cpuSeries,
+        CpuSeries.E2_MEDIUM.cpuSeries).contains(cpuSeries)) {
+      return String.format("Incorrect cpu type: %s", cpuSeries);
```
Throw an error instead

averikitsch commented on compute/cloud-client/src/main/java/compute/custommachinetype/CreateWithHelper.java:
```diff
@@ -0,0 +1,287 @@
+    CPUSeries_N2D(
+        new TypeLimits(new int[]{2, 4, 8, 16, 32, 48, 64, 80, 96}, 512, 8192, true, gbToMb(768))),
+    CPUSeries_N1(
+        new TypeLimits(concat(new int[]{1}, range(2, 97, 2)), 922, 6656, true, gbToMb(624)));
+
+    private final TypeLimits typeLimits;
+
+    Limits(TypeLimits typeLimits) {
+      this.typeLimits = typeLimits;
+    }
+
+    public TypeLimits getTypeLimits() {
+      return typeLimits;
+    }
+  }
+
+  static ImmutableMap<String, Limits> typeLimitsMap = ImmutableMap.<String, Limits>builder()
+      .put("N1", Limits.CPUSeries_N1)
+      .put("N2", Limits.CPUSeries_N2)
+      .put("N2D", Limits.CPUSeries_N2D)
+      .put("E2", Limits.CPUSeries_E2)
+      .put("E2_MICRO", Limits.CPUSeries_E2MICRO)
+      .put("E2_SMALL", Limits.CPUSeries_E2SMALL)
+      .put("E2_MEDIUM", Limits.CPUSeries_E2SMALL)
+      .build();
+
+  static int[] range(int start, int stop, int step) {
+    return IntStream.range(start, stop).filter(x -> (x - start) % step == 0).toArray();
+  }
+
+  static int gbToMb(int value) {
+    return value << 10;
+  }
+
+  static int[] concat(int[] a, int[] b) {
+    int[] result = new int[a.length + b.length];
+    System.arraycopy(a, 0, result, 0, a.length);
+    System.arraycopy(b, 0, result, a.length, b.length);
+    return result;
+  }
```
We shouldn't duplicate across samples

Sita04 replied:
This helped is crucial to demonstrate in each snippet. Hence, duplicated this across 3 snippets. 
The same procedure has been followed for Python and Go samples as well.

Please advice if this can be made more simpler.

averikitsch commented on compute/cloud-client/src/main/java/compute/custommachinetype/CreateWithHelper.java:
```diff
@@ -0,0 +1,287 @@
+      return Instance.getDefaultInstance();
+    }
+
+    try (InstancesClient instancesClient = InstancesClient.create()) {
```
client initialization needs a comment

Sita04 replied:
Done. Thanks.

averikitsch commented on compute/cloud-client/src/main/java/compute/custommachinetype/CreateWithHelper.java:
```diff
@@ -0,0 +1,287 @@
+    // Check for error string.
+    if (!machineTypeResponse.startsWith("zones/")) {
+      System.out.printf("Unable to create custom machine type string: %s%n", machineTypeResponse);
+      return Instance.getDefaultInstance();
```
Why are we returning a default instance?

Sita04 commented on compute/cloud-client/src/main/java/compute/custommachinetype/CreateCustomMachineType.java:
```diff
@@ -0,0 +1,105 @@
+
+      AttachedDisk attachedDisk = AttachedDisk.newBuilder()
+          .setInitializeParams(
+              // Describe the size and source image of the boot disk to attach to the instance.
```
Size and source image are specific to the initializeParams() and not the AttachedDisk object.
We also set other params in AttachedDisk. Hence, I would prefer to leave the comment as is?

Shabirmean replied:
Cool. Thank you for the clarification!

Sita04 commented on compute/cloud-client/src/main/java/compute/custommachinetype/CreateInstanceWithCustomSharedCore.java:
```diff
@@ -0,0 +1,283 @@
+    createInstanceWithCustomSharedCore(projectId, zone, instanceName, cpuSeries, memory);
+  }
+
+  // Construct custom machine type.
```
Modified it to: ```Construct URI for a custom machine type with a shared core.```
PTAL.

Sita04 commented on compute/cloud-client/src/main/java/compute/custommachinetype/CreateInstanceWithCustomSharedCore.java:
```diff
@@ -0,0 +1,283 @@
+    TypeLimits typeLimit = Objects.requireNonNull(
+        typeLimitsMap.get(CpuSeries.get(cpuSeries).name())).typeLimits;
+
+    // Check whether the requested parameters are allowed.
+    // Find more information about limitations of custom machine types at:
+    // https://cloud.google.com/compute/docs/general-purpose-machines#custom_machine_types
```
This is an overall comment referring to the below checks. I've modified it to be more intuitive. PTAL.,

Sita04 commented on compute/cloud-client/src/main/java/compute/custommachinetype/CreateInstanceWithCustomSharedCore.java:
```diff
@@ -0,0 +1,283 @@
+    if (typeLimit.allowedCores.length > 0 && Arrays.stream(typeLimit.allowedCores)
+        .noneMatch(x -> x == coreCount)) {
+      return String.format(
+          "Invalid number of cores requested. Allowed number of cores for %s is: %s",
```
Done. Thanks.

Sita04 commented on compute/cloud-client/src/main/java/compute/custommachinetype/CreateInstanceWithCustomSharedCore.java:
```diff
@@ -0,0 +1,283 @@
+      return String.format("zones/%s/machineTypes/%s-%s", zone, cpuSeries, memory);
+    }
+
+    if (memory > coreCount * typeLimit.maxMemPerCore) {
```
Done.

Sita04 commented on compute/cloud-client/src/main/java/compute/custommachinetype/CreateInstanceWithCustomSharedCore.java:
```diff
@@ -0,0 +1,283 @@
+  public static Instance createInstanceWithCustomSharedCore(String project, String zone,
+      String instanceName, String cpuSeries, int memory)
+      throws IOException, ExecutionException, InterruptedException, TimeoutException {
+    // Construct the machine type.
```
Done.

Sita04 commented on compute/cloud-client/src/main/java/compute/custommachinetype/CreateInstanceWithCustomSharedCore.java:
```diff
@@ -0,0 +1,283 @@
+      String instanceName, String cpuSeries, int memory)
+      throws IOException, ExecutionException, InterruptedException, TimeoutException {
+    // Construct the machine type.
+    String machineTypeResponse = customMachineTypeSharedCoreUri(zone, cpuSeries, memory);
```
Done.

Sita04 commented on compute/cloud-client/src/main/java/compute/custommachinetype/CreateInstanceWithCustomSharedCore.java:
```diff
@@ -0,0 +1,283 @@
+    }
+  }
+
+  // The limits for various CPU types are described on:
```
Modified. thanks.

Shabirmean commented on compute/cloud-client/src/main/java/compute/custommachinetype/CreateInstanceWithCustomSharedCore.java:
```diff
@@ -0,0 +1,283 @@
+
+      AttachedDisk attachedDisk = AttachedDisk.newBuilder()
+          .setInitializeParams(
+              // Describe the size and source image of the boot disk to attach to the instance.
```
Sure. Makes sense

Sita04 commented on compute/cloud-client/src/main/java/compute/custommachinetype/CreateCustomMachineType.java:
```diff
@@ -0,0 +1,105 @@
+      }
+      System.out.printf("Instance created : %s", instanceName);
+      System.out.println("Operation Status: " + response.getStatus());
+      return instancesClient.get(project, zone, instanceName);
```
Refactored the code to remove the return.

Sita04 commented on compute/cloud-client/src/main/java/compute/custommachinetype/CreateInstanceWithCustomSharedCore.java:
```diff
@@ -0,0 +1,283 @@
+
+public class CreateInstanceWithCustomSharedCore {
+
+  public enum CpuSeries {
```
The helper code is an important part of the snippet workflow. Hence, would prefer to retain it as part of the same snippet.
Wdyt?

Shabirmean commented on compute/cloud-client/src/main/java/compute/custommachinetype/CreateWithHelper.java:
```diff
@@ -0,0 +1,307 @@
+    // The amount of memory for the VM instance, in megabytes.
+    int memory = 256;
+
+    createInstanceWithCustomMachineTypeWithHelper(projectId, zone, instanceName, cpuSeries,
```
(nit) move all arguments to a anew line and in a single line

Shabirmean commented on compute/cloud-client/src/main/java/compute/custommachinetype/CreateInstanceWithCustomSharedCore.java:
```diff
@@ -0,0 +1,304 @@
+    return String.format("zones/%s/machineTypes/%s-%s-%s", zone, cpuSeries, coreCount, memory);
+  }
+
+  static ImmutableMap<String, Limits> typeLimitsMap = ImmutableMap.<String, Limits>builder()
```
Can we move this above the method where it is used close to other class variable?

Shabirmean commented on compute/cloud-client/src/main/java/compute/custommachinetype/CreateWithHelper.java:
```diff
@@ -0,0 +1,307 @@
+    // Perform the following checks to verify if the requested parameters are allowed.
+    // Find more information about limitations of custom machine types at:
+    // https://cloud.google.com/compute/docs/general-purpose-machines#custom_machine_types
+    //
```
(nit) remove the commenting on this empty line. But leave this space between the above comment and the one below

Shabirmean commented on compute/cloud-client/src/main/java/compute/custommachinetype/CreateWithHelper.java:
```diff
@@ -0,0 +1,307 @@
+      this.typeLimits = typeLimits;
+    }
+
+    public TypeLimits getTypeLimits() {
```
Do we have a precedence as to what order to place the different parts of the code? I see
- Variable initializations
- Static block initialization
- Enums
-  Inner classes
- Methods
- Main 

spread across the class. was there any specific guidance followed? If not can we structure it as:

- inner class
- enums
- class variables
- static block initialization
- main
- methods

Shabirmean replied:
This comment applies to multiple files

Sita04 replied:
Thanks for bringing this up. This is how it was supposed to be structured. But I guess java lint formatter mixed this up for me! 

Modified the format in the latest commit. PTAL.

Shabirmean commented on compute/cloud-client/src/main/java/compute/custommachinetype/HelperClass.java:
```diff
@@ -0,0 +1,269 @@
+
+    @Override
+    public String toString() {
+
```
(nit) remove extra spacing

Shabirmean commented on compute/cloud-client/src/main/java/compute/custommachinetype/HelperClass.java:
```diff
@@ -0,0 +1,269 @@
+  }
+
+  static class CustomMachineType {
+
```
(nit) remove extra spacing

Shabirmean commented on compute/cloud-client/src/main/java/compute/custommachinetype/HelperClass.java:
```diff
@@ -0,0 +1,269 @@
+      validate(cmt);
+    } catch (Error e) {
+      // Error in validation.
+      System.out.printf("Error in validation: %s", e);
```
If there is an error do we still return the same `cmt` object?

Sita04 replied:
Nice catch. Updated it.
Thanks.

Shabirmean reviewed (commented):
Thank you for addressing the previous comments. I have left some minor suggestions which are just nits.

However, I had a brad question about the organization of the classes. PTAL and let me know what are your thoughts

Sita04 commented:
Thanks @Shabirmean for the review. Addressed all your comments.
The class structure was supposed to be in the format you suggested. Looks like the auto code linter had other plans :) Thanks for bringing it up!

Shabirmean reviewed (approved):
LGTM!

Thank you for addressing the comments
